File: OJ/leet456/leet456.py
# leet456

from typing import List


class Solution:
        
    def find132pattern1(self, A: List[int]) -> bool:
        # TLE, O(N^2), 
        for i in range(len(A)):
            curMax = A[i]
            for j in range(i+1, len(A)):
                if A[i] < A[j]:
                    if A[j] < curMax: return True 
                    curMax = A[j] 
        return False

    def find132pattern(self, A: List[int]) -> bool:
        # O(N) 栈
        if len(A) < 3: return False

        minA = [A[0]]
        for i in range(1, len(A)):
            minA.append(min(minA[-1], A[i]))

        # 维护一个递减栈
        # [0..j) aj [j+1,..n)
        # M1         M2
        # 要求 M1 < M2 < aj 
        # 要点：
        # 到第 j 个数时，栈中小于 min(A[0..j]) 的数都是无用的
        S = []
        for j in range(len(A)-1, -1, -1):
            if A[j] > minA[j]:
                while S and S[-1] <= minA[j]: S.pop()
                if S and S[-1] < A[j]: return True
                S.append(A[j])
        return False


print(Solution().find132pattern([1, 2, 3, 4]))
print(Solution().find132pattern([3, 1, 4, 2]))
print(Solution().find132pattern([-1, 3, 2, 0]))
print(Solution().find132pattern([1,0,1,-4,-3]))
print(Solution().find132pattern([3,5,0,3,4]))

File: OJ/leet456/leet456.cpp
/*
leet456


*/


#include <iostream>
#include <tuple>
#include <vector>
#include <queue>
#include <stack>
#include <string>
#include <set>
#include <map>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>
#include <numeric>
#include <cassert>
#include <random>
#include <functional>
#include <utility>
#include "../utils/LeetCpp.utils.hpp"

using namespace std;

class Solution {
public:
    bool find132pattern(const vector<int>& A) {
        // 枚举 1 O(N)
        int n = A.size();
        stack<int> S; 
        auto max2 = INT_MIN;

        for (int i = n - 1; i >= 0; i--)
        {
            auto x = A[i];
            if (x < max2) return true; // x 作为 max1： 成功
            while (S.size() and S.top() < x) {
                max2 = S.top(); S.pop(); // 我们希望 max2 尽可能大
            }
            if (x > max2) S.push(x); // x 作为 max3
        }
        return false;
    }
};

int main(int argc, char const *argv[])
{
    Solution sol;   
    cout << boolalpha;
    cout << sol.find132pattern({1,2,3,4}) << endl;
    cout << sol.find132pattern({3,1,4,2}) << endl;
    cout << sol.find132pattern({-1,3,2,0}) << endl;
    return 0;
}
